File: app/src/main/java/wangdaye/com/geometricweather/basic/GeoDialog.java
package wangdaye.com.geometricweather.basic;

import android.content.Context;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.LifecycleEventObserver;

import wangdaye.com.geometricweather.R;

public abstract class GeoDialog extends DialogFragment {

    private boolean mForeground = false;

    public static void injectStyle(DialogFragment f) {
        f.getLifecycle().addObserver((LifecycleEventObserver) (source, event) -> {
            switch (event) {
                case ON_CREATE:
                    f.setStyle(STYLE_NO_TITLE, android.R.style.Theme_DeviceDefault_Dialog_MinWidth);
                    break;

                case ON_START:
                    f.requireDialog().getWindow().setBackgroundDrawableResource(R.drawable.dialog_background);
                    break;
            }
        });
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        injectStyle(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        mForeground = true;
        ((GeoActivity) requireActivity()).setTopDialog(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        mForeground = false;
        ((GeoActivity) requireActivity()).checkToCleanTopDialog(this);
    }

    public abstract View getSnackbarContainer();

    public boolean isForeground() {
        return mForeground;
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/common/basic/GeoDialog.kt
package wangdaye.com.geometricweather.common.basic

import android.content.Context
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver

abstract class GeoDialog : DialogFragment() {

    companion object {
        fun injectStyle(f: DialogFragment) {
            f.lifecycle.addObserver(LifecycleEventObserver { _, event ->
                when (event) {
                    Lifecycle.Event.ON_CREATE -> if (f.view != null) {
                        f.setStyle(STYLE_NO_TITLE, android.R.style.Theme_DeviceDefault_Dialog_MinWidth)
                    }
                    Lifecycle.Event.ON_START -> if (f.view != null) {
                        f.requireDialog().window!!.setBackgroundDrawableResource(wangdaye.com.geometricweather.R.drawable.dialog_background)
                    }
                    else -> {}
                }
            })
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        injectStyle(this)
    }
}